teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,544 @@
+```
+{
+        "uri": "<URI>",
+        "digest": { /* DigestSet */ }
+ }
+```
+The other specific requirements of these fields can be found in [in-toto attestation spec](https://github.com/in-toto/attestation/tree/main/spec). Another thing to be aware is that For `digest`, it has a key for the digest type and has a value for that key. Git, Mercurial, Bazaar and Subversion have full VCS location support. CVS, Perforce are providing this VCS location as plain URLs. For VCS, the list of VCS that the community are looking into is: Git, Mercurial, Subversion, Bazaar, CVS, Perforce. Currently Chains only supports Git revisions for VCS and OCI images as signable targets for inputs.
+
+For the `outputs`, we are looking at the following support, as these are the artifact types supported in Artifact Registries. However, the list should definitely be expanded as this doc is spreading the discussions in the communities to see which should be prioritized. Currently, we have the following under considerations: Python, Maven, Go, NodeJS and OCI images. Note that OCI images are the ones we currently support.
+
+#### Concrete Use Cases
+All below examples will be generating in-toto attestation.
+ 1. Use Git commits as sources, and generates in-toto provenance for a TaskRun that builds an image.
+   
+    A Task example `without` structured results and params support can be like:
+    ```
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-GIT_COMMIT
+          type: string
+          description: git commit sha
+        - name: CHAINS-GIT_URL
+          type: string
+          description: git commit url
+        - name: FOO_IMAGE
+          type: string
+        - description: artifact image
+      ...
+      results:
+        - name: IMAGE_DIGEST
+          description: digest of image
+    ```
+    A Task example `with` structured results and params support will be like:
+    ```
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-VCS
+          type: object
+          description: |
+            The source distribution
+            * uri: resource uri of the git commit
+            * digest: revision digest
+            * alg: digest algorithm: sha256 etc.
+            * src_type: source VCS type: git, hg, etc.
+          properties:
+          src_type:
+            type: string
+          uri:
+            type: string
+          alg:
+            type: string
+          digest:
+            type: string
+      ...
+      results:
+        - name: CHAINS-OCI_IMAGE
+          type: object
+          description: |
+            The source distribution
+            * name: name of the built image
+            * digest: image digest
+            * alg: digest algorithm: sha256 etc.
+          properties:
+          name:
+            type: string
+          digest:
+            type: string
+          alg:
```

Comment:
It's not clear that this is the algorithm of the digest. Maybe `digest_ago` is more specific? Alternatively `digest` could be a specially formatted string that includes both information, e.g. `<algo>:<digest>`.

Reply:
I changed the name to `digest_ago`. I'm still keeping them separate to give more flexibilities.